api/types/matchers_aws.go

```diff
@@ -145,10 +150,24 @@ func (m *AWSMatcher) CheckAndSetDefaults() error {
 	}
 
 	if len(m.Regions) == 0 {
-		return trace.BadParameter("discovery service requires at least one region")
+		m.Regions = []string{Wildcard}
 	}
 
 	for _, region := range m.Regions {
+		if region == Wildcard {
+			if len(m.Regions) > 1 {
+				return trace.BadParameter("when using %q as region, no other regions can be specified", Wildcard)
+			}
+
+			// Only EC2 supports discovering regions.
+			// TODO(marco): add support for other resources types.
+			if len(m.Types) > 1 || m.Types[0] != AWSMatcherEC2 {
+				return trace.BadParameter("only EC2 resource discovery supports discovering all regions, " +
+					"enumerate all the regions or create a separate matcher for discovering EC2 resources")
+			}
+			break
```

Comment:
Should we move this out of the CheckAndSetDefaults and validate this property when creating/updating the resource. This might save us a big headache when we'll need to add region discovery to other things later and older versions will get angry.

(this might not be a big issue with the cache changes anymore)